torch_xla/csrc/runtime/ifrt_computation_client.cc

```diff
@@ -0,0 +1,678 @@
+  builder.SetSharding(xla::HloSharding::Replicate().ToProto());
+  xla::XlaOp scalar_zero_op = xla::ConvertElementType(
+      xla::ConstantR0(&builder, 0), shape.element_type());
+  xla::XlaOp y = xla::Add(x, scalar_zero_op);
+  auto instruction = XlaBuilderFriend::GetInstruction(y);
+  *instruction->mutable_sharding() = xla::HloSharding::Replicate().ToProto();
+
+  xla::XlaComputation computation =
+      ConsumeValue(builder.Build(/*remove_dynamic_dimensions=*/false));
+  xla::ProgramShape program_shape = ConsumeValue(computation.GetProgramShape());
+
+  std::string device = GetDefaultDevice();
+  std::vector<torch_xla::runtime::ComputationClient::CompileInstance> instances;
+  instances.push_back({std::move(computation), device,
+                       GetCompilationDevices(device, {}), &shape,
+                       /*should_wrap_parameter=*/false,
+                       /*is_sharded=*/true,
+                       /*allow_spmd_sharding_propagation_to_output=*/false});
+  std::vector<
+      std::shared_ptr<torch_xla::runtime::ComputationClient::Computation>>
+      computations = Compile(std::move(instances));
+
+  XLA_CHECK_EQ(handle->buffer->sharding().devices().size(),
+               GetLocalDevices().size());
+
+  torch_xla::runtime::ComputationClient::ExecuteReplicatedOptions
+      execute_options;
+
+  auto sharded_results = ExecuteReplicated(*computations.front(), {{handle}},
+                                           GetLocalDevices(), execute_options);
+  auto replicated_output =
+      std::dynamic_pointer_cast<IfrtData>(sharded_results[0])
+          ->buffer->FullyReplicatedShard(
+              xla::ifrt::ArrayCopySemantics::kAlwaysCopy);
+  // TODO: sanity check outputs
+  return *replicated_output;
+}
+
+std::vector<xla::Literal> IfrtComputationClient::TransferFromServer(
+    absl::Span<const DataPtr> handles) {
+  metrics::TimedSection timed(TransferFromServerMetric());
+  tsl::profiler::TraceMe activity("IfrtComputationClient::TransferFromServer",
+                                  tsl::profiler::TraceMeLevel::kInfo);
+  std::vector<xla::Literal> literals;
+  literals.reserve(handles.size());
+  int64_t total_size = 0;
+  for (auto handle : handles) {
+    // Use XLA replication to reassemble the sharded data. If input handle
+    // is not sharded, then it is a no-op.
+    auto ifrt_data = std::dynamic_pointer_cast<IfrtData>(handle);
+    tsl::RCReference<xla::ifrt::Array> replicated_array =
+        ReplicateShardedData(ifrt_data);
+
+    // TODO: handle dynamic shapes
+    auto& literal = literals.emplace_back(
+        xla::ShapeUtil::DeviceShapeToHostShape(ifrt_data->shape()));
+    std::vector<int64_t> byte_strides(literal.shape().dimensions_size());
+    XLA_CHECK_OK(xla::ShapeUtil::ByteStrides(literal.shape(),
+                                             absl::MakeSpan(byte_strides)));
+    XLA_CHECK_OK(
+        replicated_array
+            ->CopyToHostBuffer(literal.untyped_data(), byte_strides,
```

Comment:
Follow-up: we'll need to parallelize this if possible for async checkpointing performance.

Reply:
Agreed